peep.c

```diff
@@ -3590,41 +3590,110 @@ Perl_rpeep(pTHX_ OP *o)
         case OP_COND_EXPR:
             if (o->op_type == OP_COND_EXPR) {
                 OP *stub = cLOGOP->op_other;
+                OP *trueop  = OpSIBLING( cLOGOP->op_first );
+                OP *falseop = OpSIBLING(trueop);
+
                 /* Is there an empty "if" block or ternary true branch?
                    If so, optimise away the OP_STUB if safe to do so. */
                 if (stub->op_type == OP_STUB &&
                     ((stub->op_flags & OPf_WANT) != OPf_WANT_SCALAR)
                 ) {
-                    OP *trueop = OpSIBLING( cLOGOP->op_first );
-
-                    assert((stub == trueop ) || (OP_TYPE_IS(trueop, OP_SCOPE) &&
-                     ((stub == cUNOPx(trueop)->op_first)) && !OpSIBLING(stub))
-                    );
-                    assert(!(stub->op_flags & OPf_KIDS));
-
-                    cLOGOP->op_other = (stub->op_next == trueop) ?
-                                  stub->op_next->op_next  :
-                                  stub->op_next;
-
-                    op_sibling_splice(o, cLOGOP->op_first, 1, NULL);
-
-                    if (stub != trueop) op_free(stub);
-                    op_free(trueop);
-                } else
+                    if (stub == trueop) {
+                        /* This is very unlikely:
+                         *     cond_expr
+                         *         -condition-
+                         *         stub
+                         *         -else-
+                         */
+                        assert(!(stub->op_flags & OPf_KIDS));
+                        cLOGOP->op_other = stub->op_next;
+                        op_sibling_splice(o, cLOGOP->op_first, 1, NULL);
+                        op_free(stub);
+                        break;
+                    } else if (OP_TYPE_IS(trueop, OP_SCOPE) &&
+                               (stub == cUNOPx(trueop)->op_first) ) {
+                        assert(!(stub->op_flags & OPf_KIDS));
+
+                        OP *stubsib = OpSIBLING(stub);
+                        if (!stubsib) {
+                        /*     cond_expr
+                         *         -condition-
+                         *         scope
+                         *             stub
+                         *         -else-
+                         */
+                            cLOGOP->op_other = trueop->op_next;
+                            op_sibling_splice(o, cLOGOP->op_first, 1, NULL);
+                            op_free(stub);
+                            op_free(trueop);
```

Comment:
I had forgotten that `op_free` will free kids, so this should be just `op_free(trueop)` before being merged.